docs/sat-solver.md

```diff
@@ -0,0 +1,439 @@
+
+Installing <img src="https://render.githubusercontent.com/render/math?math=p_{1}"> implies
+installing <img src="https://render.githubusercontent.com/render/math?math=x_{1}">.
+Although <img src="https://render.githubusercontent.com/render/math?math=x_{1}"> may be
+installed without <img src="https://render.githubusercontent.com/render/math?math=p_{1}">.
+
+
+### Depends that can be satisfied by multiple versions
+
+If <img src="https://render.githubusercontent.com/render/math?math=p_{1}">
+depends on multiple versions of _x_, for example,
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}"> and
+<img src="https://render.githubusercontent.com/render/math?math=x_{2}">:
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1} %2B x_{2} - p_{1} \geq 0">
+
+
+### Conflicts
+
+If <img src="https://render.githubusercontent.com/render/math?math=y_{2}"> conflicts with
+<img src="https://render.githubusercontent.com/render/math?math=x_{1}">:
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1} %2B y_{2} \leq 1">
+
+
+### Provides
+
+If <img src="https://render.githubusercontent.com/render/math?math=x_{1}">
+provides
+<img src="https://render.githubusercontent.com/render/math?math=b_{1}">:
+
+<img src="https://render.githubusercontent.com/render/math?math=x_{1} \geq 1">,
+<img src="https://render.githubusercontent.com/render/math?math=b_{1} \geq 1">
+
+### Optional-Depends
+
+If <img src="https://render.githubusercontent.com/render/math?math=x_{1}">
+optional-depends on
+<img src="https://render.githubusercontent.com/render/math?math=a_{1}, b_{1}">: 
+
+<img src="https://render.githubusercontent.com/render/math?math=a_{1} %2B b_{1} \geq 0">, 
+together with
+<img src="https://render.githubusercontent.com/render/math?math=x_{1} \geq 1">
+
+## Objective function
+
+The quality of the solution depends on the definition of the objective function.
+We can use several criteria when optimizing the objective function.
+
+### Minimize package removal
+
+Given  <img src="https://render.githubusercontent.com/render/math?math=PI'_{1} \dots PI'_{n}">
+as packages already installed:
+
+<img src="https://render.githubusercontent.com/render/math?math=f_{1} = min( (1 - PI'_{1}) %2B \dots %2B (1 - PI'_{n}))">
+
+The solver will try to set <img src="https://render.githubusercontent.com/render/math?math=PI_{i}">
+to 1, which implies not removing the package.
+
+### Minimize the number of installed packages
+
+Given <img src="https://render.githubusercontent.com/render/math?math=P_{1} \dots P_{n}">
+as the new packages targetted to be installed (either existent or new), the
+objective function will be:
+
+<img src="https://render.githubusercontent.com/render/math?math=f_{2}(P) = min(P_{1} %2B \dots %2B P_{n})">
+
+### Maximizing the freshness of packages
+
+When the objective is to install the most recent versions, independently of what
+is already installed.
```

Comment:
Ideally we would keep the existing version running rather than upgrading. This was requested from the Rancher folks who would be one of the users of the project. Possibly upgrading if a flag is provided.

Reply:
Sounds good. Do we want to not bump *any* version, or do we want to upgrade to minor and patch versions?

Reply:
We _do not_ want to upgrade any version. The reason is that they may have passed custom values when installing it. An upgrade could cause those custom values to be lost on upgrade. It is not safe to do an upgrade.

Reply:
That's consistent with the current implementation.
The docs are listing that possibility (and others), in case that in the future we want to implement an "upgrade all to major/minor/patch" command, or a "check cluster", for example.